### PR TITLE
Revised reduce interface - issue 121

### PR DESCRIPTION
**Pull request by Drioastorga**
Modified reduce interface.

Changes:
- Reduce interface versions targeted to iterable outputs have been removed. 
- Modified interface. Now the reduce supports any reduce operation (lamda or standard binary operators) and receive an initial (identity) value for those operations that cannot be deduced. Alternatively, there is a new interface that does not recieve any initial value but computes the identity value if the input types can be casted to boolean. Note that the reduce interface only supports homogeneous reduce operations, i.e T x = op(T y, Y z) being T the container value type.
- Updated test to match with the new interface.

### CODE REVIEW

File: include/native/reduce.h [hunk 1]
```diff
@@ -6,7 +6,13 @@
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
-*
+*te < typename InputIt, typename ReduceOperator>
```
Comment (jdgarciauc3m) [inline include/native/reduce.h]: I imagine this is a mistake. What's the point of this in the middle of the license?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Anything that can be forwarded sould be a forwarding reference argument.

In general, make all function object arguments forwarding references.

File: include/native/reduce.h (hunk 1 above)
Comment (Drioastorga) [reply include/native/reduce.h]: It was a mistake. I have removed those lines.

**Review by Drioastorga — COMMENTED** [review]
I have modified the reduce interface to pass the function objects as forwarding references.

File: include/native/mapreduce.h [hunk 2]
```diff
@@ -31,7 +31,7 @@ template < typename InputIt, typename OutputIt, typename MapFunc, typename Reduc
     p.register_thread();
     while( first != last ) {
        auto mapresult = map(*first, inputs ... );
-       reduce(p, mapresult.begin(), mapresult.end(), *firstOut, op);
+       *firstOut = reduce(p, mapresult.begin(), mapresult.end(), op);
```
Comment (jdgarciauc3m) [inline include/native/mapreduce.h]: You need to forward op here

File: include/native/reduce.h [hunk 3]
```diff
@@ -25,154 +25,65 @@
...
-
-template < typename InputIt, typename Output, typename ReduceOperator>
- typename std::enable_if<!is_iterator<Output>::value, void>::type 
-reduce(parallel_execution_native &p, InputIt first, InputIt last, Output & firstOut, ReduceOperator op) {
-
-    typename ReduceOperator::result_type identityVal = !op(false,true);
 
+template < typename InputIt, typename ReduceOperator>
+typename std::iterator_traits<InputIt>::value_type
+reduce(parallel_execution_native &p, InputIt first, InputIt last, typename std::iterator_traits<InputIt>::value_type init, ReduceOperator && op){
```
Comment (jdgarciauc3m) [inline include/native/reduce.h]: In all grppi::reduce the operation type should be renamed to `Combiner` and the argument to `combine_op`.

In all grppi::map_reduce the operation type should be renamed to `Transformer` and the argument to `transform_op`.

File: include/native/reduce.h [hunk 4]
```diff
@@ -25,154 +25,65 @@
...
+   auto identityVal = !op(false,true);
+   return reduce(p, first, last, identityVal, std::forward<ReduceOperator>(op));
 }
 
 
 
-
-template < typename InputIt, typename ReduceOperator>
- typename ReduceOperator::result_type
+//typename ReduceOperator::result_type
```
Comment (jdgarciauc3m) [inline include/native/reduce.h]: What are this comments for?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
I made some comments. Try to address them generally.

File: include/native/reduce.h (hunk 4 above)
Comment (Drioastorga) [reply include/native/reduce.h]: These comments have been removed.

File: include/native/reduce.h (hunk 3 above)
Comment (Drioastorga) [reply include/native/reduce.h]: Type and function object names have been acordingly modified in reduce and map_reduce interfaces.

File: include/native/mapreduce.h (hunk 2 above)
Comment (Drioastorga) [reply include/native/mapreduce.h]: Fixed.

**Review by Drioastorga — COMMENTED** [review]
Types and function objects names have been modified.

File: include/native/mapreduce.h [hunk 5]
```diff
@@ -60,19 +60,19 @@ template <typename InputIt, typename MapFunc, class T, typename ReduceOperator>
             // Register thread
             p.register_thread();
 
-            partialOuts[i] = MapReduce(sequential_execution{}, begin, last, std::forward<MapFunc>(map), op);
+            partialOuts[i] = MapReduce(sequential_execution{}, begin, last, std::forward<Transformer>(transform_op), std::forward<Combiner>(combine_op));
```
Comment (jdgarciauc3m) [inline include/native/mapreduce.h:63]: What is MapReduce calling to?
Is this really used?

File: include/native/mapreduce.h [hunk 6]
```diff
@@ -60,19 +60,19 @@ template <typename InputIt, typename MapFunc, class T, typename ReduceOperator>
...
          begin, end, out)
        );
     }
-    partialOuts[0] = MapReduce(sequential_execution{}, begin, last, std::forward<MapFunc>(map), op);
+    partialOuts[0] = MapReduce(sequential_execution{}, begin, last, std::forward<Transformer>(transform_op), std::forward<Combiner>(combine_op));
     for(auto task = tasks.begin();task != tasks.end();task++){    
        (*task).join();
     }
-    Reduce(sequential_execution{}, partialOuts.begin(), partialOuts.end(), out, op);
+    Reduce(sequential_execution{}, partialOuts.begin(), partialOuts.end(), out, std::forward<Combiner>(combine_op));
```
Comment (jdgarciauc3m) [inline include/native/mapreduce.h:75]: What is this Reduce() calling to?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Address comments.

Besides building of this branch fails

[ 85%] Building CXX object legacy_utest/CMakeFiles/divideandconquer1_GT.dir/divideandconquer1_GT.cpp.o
/home/jdgarcia/github-repos/arcosuc3m/grppi/legacy_utest/reduce2_GT.cpp: In instantiation of ‘int reduce_example2(auto:1&) [with auto:1 = grppi::sequential_execution]’:
/home/jdgarcia/github-repos/arcosuc3m/grppi/legacy_utest/reduce2_GT.cpp:65:5:   required from here
/home/jdgarcia/github-repos/arcosuc3m/grppi/legacy_utest/reduce2_GT.cpp:53:12: error: no matching function for call to ‘reduce(grppi::sequential_execution&, std::vector<std::vector<int> >::iterator, std::vector<std::vector<int> >::iterator, std::vector<int>::iterator, reduce_example2(auto:1&) [with auto:1 = grppi::sequential_execution]::<lambda(auto:2&, auto:3&)>)’
      reduce(p, in.begin(), in.end(), out.begin(), [&](auto & in, auto & out){
      ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
              reduce(p, in.begin(), in.end(), out,  std::plus<int>() );
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          }
          ~  
      );
      ~      
In file included from /home/jdgarcia/github-repos/arcosuc3m/grppi/include/reduce.h:28:0,

**Comment by Drioastorga** [comment]
The changes of the mapreduce should be addressed for the issue #122. I will take these comments into account for the aforementioned issue.

**Review by jdgarciauc3m — APPROVED** [review]
OK

**Comment by jdgarciauc3m** [comment]
Solves issue #121

**Pull request merged**